### PR TITLE
Add std::ops::Index support for infering

### LINKED ISSUE #2534: Unable to inference a result type of an indexing operation
Hi everyone, I was under impression that type inference involving arrays should work, but it is behaving weirdly:
```rust
fn foo(bar: [usize; 2]) -> usize {
    let baz = bar[1];
    baz
}

fn foo2(bar: [usize; 2]) -> usize {
    let baz = bar[1];
    return baz;
}

fn foo3(bar: [usize; 2]) {
    let baz = bar[1];
    println!("{}", baz);
}

fn main() {
    let foo4 = |bar: [usize; 2]| -> usize {
        let baz = bar[1];
        baz
    };

    let foo5 = |bar: [usize; 2]| -> usize {
        let baz = bar[1];
        return baz;
    };

    println!(
        "{} {} {} {}",
        foo([1, 2]),
        foo2([1, 2]),
        foo4([1, 2]),
        foo5([1, 2])
    );
    foo3([1, 2]);
}
```

The type inferred for `baz` is `usize` for `foo1` and `foo2` (good), `{unknown}` for `foo3` and `foo4` (I would expect that type inference is not too hard there) and `()` for `foo5` (plain wrong!).

Tested with the latest master (https://github.com/rust-analyzer/rust-analyzer/commit/4444192b05c107a40a5a05ea3c9091ad8f8cbbcc)

Edit: there are two separate issues: type in foo3 and foo4 is not inferred because it is an indexing operation and is inferred incorrectly in foo5 because of the `return` statement in the closure (#2547 )

### PR DESCRIPTION
**Pull request by edwin0cheng**
see also #2534

Seem like this can't fix #2534 for this case:

```rust
fn foo3(bar: [usize; 2]) {
    let baz = bar[1];   // <--- baz is still unknown ?
    println!("{}", baz);
}
```

### CODE REVIEW

**Comment by edwin0cheng** [comment]
cc @flodiebold

**Comment by edwin0cheng** [comment]
Seem like this PR is not enough to prove all trait associated types for `[T]` impl for `std::ops::Index` ? 
```rust
impl<T, I> ops::Index<I> for [T]
    where I: SliceIndex<[T]>
{
    type Output = I::Output;

    #[inline]
    fn index(&self, index: I) -> &I::Output {
        index.index(self)
    }
}
```

**Comment by kiljacken** [comment]
I have some input on the issue of the example and a little on the new
helper func, but currently stuck in public transit so it will be an hour or
so :)

File: crates/ra_hir_ty/src/infer.rs [hunk 1]
```diff
@@ -363,14 +363,28 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
...
+    fn resolve_associated_type_with_params(
+        &mut self,
+        inner_ty: Ty,
+        assoc_ty: Option<TypeAliasId>,
+        params: &[Ty],
+    ) -> Ty {
         match assoc_ty {
             Some(res_assoc_ty) => {
                 let ty = self.table.new_type_var();
+                let mut builder = Substs::builder(1 + params.len()).push(inner_ty);
```
Comment (kiljacken) [inline crates/ra_hir_ty/src/infer.rs]: There's a nice helper to ensure the right amount of params is supplied (not sure if the code compiles, but should give the general idea):
```suggestion
                let mut builder = Substs::build_for_def(&self.db, res_assoc_ty).push(inner_ty);
```

**Comment by kiljacken** [comment]
As for the issue with the example above, the problem is two-fold:
1. Some amount of autoderef is needed on the type being indexed, but I never quite managed to figure out how that works. @matklad / @flodiebold are one of you able to give a short intro to how our autoderef stuff works?
2. Index is only implemented for slices, but arrays with size coerces to slices in rust, so we need to handle that coercion.

**Comment by kiljacken** [comment]
Also just thought of the following: Once this is in place it can be trivially copied and adapted to handle custom BinOp implementations!
Maybe it would make sense to handle indexing as a binary op in hir? They seems to need the same kind of autoderef and inference handling.

**Comment by matklad** [comment]
r? @flodiebold 

>Maybe it would make sense to handle indexing as a binary op in hir? 

See https://github.com/rust-analyzer/rust-analyzer/pull/1694#issue-308298275. I think index is different from other binops in this case, but I might be mistaken.

**Comment by kiljacken** [comment]
> See [#1694 (comment)](https://github.com/rust-analyzer/rust-analyzer/pull/1694#issue-308298275). I think index is different from other binops in this case, but I might be mistaken.

Ahh you've had the same thought already :) Right, that conclusion does make some sense, keeping it seperate seems like it will remove a layer of enum matching when handling `&mut` on index expressions at some point.

File: crates/ra_hir_ty/src/infer.rs (hunk 1 above)
Comment (edwin0cheng) [reply crates/ra_hir_ty/src/infer.rs]: Oh, nice. I do not know about that helper.

**Comment by edwin0cheng** [comment]
> 
> 
> As for the issue with the example above, the problem is two-fold:
> 
>     1. Some amount of autoderef is needed on the type being indexed, but I never quite managed to figure out how that works. @matklad / @flodiebold are one of you able to give a short intro to how our autoderef stuff works?
> 
>     2. Index is only implemented for slices, but arrays with size coerces to slices in rust, so we need to handle that coercion.

@kiljacken  And I found that the following case do not work too and I don't know how to add that `IndexWithOutput` associate type information to the solver...

```rust
struct Bar;
struct Foo;

trait IndexWithOutput {
    type Output;
}

impl IndexWithOutput for u32 {
    type Output = Foo;
}

impl<T:IndexWithOutput> std::ops::Index<T> for Bar {
    type Output = T::Output;
}

fn test() {
    let a = Bar;
    let b = a[1];
    b<|>;
}
```

**Comment by flodiebold** [comment]
> And I found that the following case do not work too and I don't know how to add that `IndexWithOutput` associate type information to the solver...

Without investigating in detail, I'd suspect https://github.com/rust-lang/chalk/issues/234 for this. As I mentioned in #2454, that bug sadly prevents associated type normalization from working in a lot of non-trivial cases. Of course it could very well be something else, but I'm currently a bit reluctant to investigate associated type issues deeply because of it...

> Some amount of autoderef is needed on the type being indexed, but I never quite managed to figure out how that works. @matklad / @flodiebold are one of you able to give a short intro to how our autoderef stuff works?

Hmm, actually this probably also does autoref, right? Should it be handled like method resolution? (It took quite a bit of care to get the order in which candidates are considered right there...) I haven't actually checked how rustc handles this...

As for autoderef, basically you just call `autoderef` and get an iterator of successively derefed types. The only complication is that you have to canonicalize the type (which basically means getting rid of inference variables so we can properly cache the trait queries) and pass the correct environment (i.e. set of where clauses). It should work exactly the same as e.g. for [field accesses](https://github.com/rust-analyzer/rust-analyzer/blob/08c6933104baca84fd4135a76cdc7daf60a0c631/crates/ra_hir_ty/src/infer/expr.rs#L238) though.

File: crates/ra_hir_ty/src/infer.rs [hunk 2]
```diff
@@ -363,14 +363,28 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
...
+        assoc_ty: Option<TypeAliasId>,
+        params: &[Ty],
+    ) -> Ty {
         match assoc_ty {
             Some(res_assoc_ty) => {
                 let ty = self.table.new_type_var();
+                let mut builder = Substs::build_for_def(self.db, res_assoc_ty).push(inner_ty);
+                for ty in params {
+                    builder = builder.push(ty.clone());
+                }
```
Comment (flodiebold) [inline crates/ra_hir_ty/src/infer.rs]: this is `builder.fill(params.iter().cloned())`, right?

Comment (flodiebold) [reply crates/ra_hir_ty/src/infer.rs]: except yours will check that the number of provided types matches exactly... maybe we need a `fill_exact`

**Comment by flodiebold** [comment]
> @kiljacken And I found that the following case do not work too and I don't know how to add that `IndexWithOutput` associate type information to the solver...

Actually, this might be a really complicated case... The `1` is an unresolved integer type variable when we're trying to infer the slice operation, so we'd need to somehow infer it's `usize`, but I don't really know how either Chalk or we would :thinking:

**Comment by kiljacken** [comment]
> Actually, this might be a really complicated case... The `1` is an unresolved integer type variable when we're trying to infer the slice operation, so we'd need to somehow infer it's `usize`, but I don't really know how either Chalk or we would thinking

I just added it as an expectation on the infer of the inner expr in my initial attempt at this, that seemed to do the trick, although I'm not quite sure if it was "correct".

**Comment by flodiebold** [comment]
> I just added it as an expectation on the infer of the inner expr in my initial attempt at this, that seemed to do the trick, although I'm not quite sure if it was "correct".

If you mean adding an expectation of `usize`, I don't think that works -- e.g. in @edwin0cheng's example rustc infers `u32`: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=42a900cdc14d79df4e51c4b2033c3441

**Comment by kiljacken** [comment]
> If you mean adding an expectation of `usize`, I don't think that works -- e.g. in @edwin0cheng's example rustc infers `u32`: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=42a900cdc14d79df4e51c4b2033c3441

Oh right, hadn't considered that. That does indeed seem like a great can of worms

**Comment by edwin0cheng** [comment]
I just checked, it do not works even without `usize`:

```rust
struct Bar;
struct Foo;
struct Baz;

trait IndexWithOutput {
    type Output;
}

impl IndexWithOutput for Baz {
    type Output = Foo;
}

impl<T:IndexWithOutput> std::ops::Index<T> for Bar {
    type Output = T::Output;
    
    fn index(&self, _index: T) -> &Self::Output { unimplemented!() }    
}

fn main() {
    let a = Bar;
    let c = Baz;
    let b = a[c];  // b is unknown
}
```

**Comment by flodiebold** [comment]
Yeah, I don't think we'll get this working right now, but we should still merge this. Although I guess we could detect when the parameter is an int variable and try using `usize` in that case... :thinking:

**Comment by edwin0cheng** [comment]
@flodiebold, indeed. I just opened another issue to check part of it.

**Comment by kiljacken** [comment]
The [librustc_typeck source](https://github.com/rust-lang/rust/blob/0de96d37fbcc54978458c18f5067cd9817669bc8/src/librustc_typeck/check/mod.rs#L3352) seems to point some light on it being a bit different from method handling, and quite an ordeal!

File: crates/ra_hir_ty/src/infer.rs (hunk 2 above)
Comment (edwin0cheng) [reply crates/ra_hir_ty/src/infer.rs]: I changed it to: 
```rust
let builder = Substs::build_for_def(self.db, res_assoc_ty)
    .push(inner_ty)
    .fill(params.iter().cloned());
```

**Comment by flodiebold** [comment]
@edwin0cheng I'd be fine with r+ing this, unless you first want to try getting autoderef working?

**Comment by edwin0cheng** [comment]
>@edwin0cheng I'd be fine with r+ing this, unless you first want to try getting autoderef working?

@flodiebold Um..  I would like to wait for https://github.com/rust-lang/chalk/issues/234 get fixed first :)

**Pull request merged**